community/index/src/main/java/org/neo4j/index/gbptree/InternalTreeLogic.java

```diff
@@ -84,15 +87,230 @@
+     * ... a.s.o
+     *
+     * Calling {@link #insert(PageCursor, StructurePropagation, Object, Object, ValueMerger, org.neo4j.index.IndexWriter.Options, long, long)}
+     * or {@link #remove(PageCursor, StructurePropagation, Object, Object, long, long)} leaves the cursor
+     * at the last updated page (tree node id) and remembers the path down the tree to where it is.
+     * Further inserts/removals will move the cursor from its current position to where the next change will
+     * take place using as few page pins as possible.
+     */
+    @SuppressWarnings( "unchecked" )
+    private Level<KEY>[] levels = new Level[0]; // grows on demand
+    private int currentLevel = -1;
+
+    /**
+     * Keeps information about one level in a path down the tree where the {@link PageCursor} is currently at.
+     *
+     * @param <KEY> type of keys in the tree.
+     */
+    private static class Level<KEY>
+    {
+        // For comparing keys
+        private final Comparator<KEY> layout;
+        // Id of the tree node id this level of the path
+        private long treeNodeId;
+
+        // Child position which was selected from parent to get to this level
+        private int childPos;
+        // Lower bound of key range this level covers
+        private final KEY lower;
+        // Whether or not the lower bound is fixed or open-ended (far left in the tree)
+        private boolean lowerIsOpenEnded;
+        // Upper bound of key range this level covers
+        private final KEY upper;
+        // Whether or not the upper bound is fixed or open-ended (far right in the tree)
+        private boolean upperIsOpenEnded;
+
+        Level( Layout<KEY,?> layout )
+        {
+            this.layout = layout;
+            this.lower = layout.newKey();
+            this.upper = layout.newKey();
+        }
+
+        /**
+         * Returns whether or not the key range of this level of the path covers the given {@code key}.
+         *
+         * @param key KEY to check.
+         * @return {@code true} if key is within the key range if this level, otherwise {@code false}.
+         */
+        boolean covers( KEY key )
+        {
+            if ( !lowerIsOpenEnded && layout.compare( key, lower ) < 0 )
+            {
+                return false;
+            }
+            if ( !upperIsOpenEnded && layout.compare( key, upper ) >= 0 )
+            {
+                return false;
+            }
+            return true;
+        }
+    }
+
     InternalTreeLogic( IdProvider idProvider, TreeNode<KEY,VALUE> bTreeNode, Layout<KEY,VALUE> layout )
     {
         this.idProvider = idProvider;
         this.bTreeNode = bTreeNode;
         this.layout = layout;
-        int maxKeyCount = max( bTreeNode.internalMaxKeyCount(), bTreeNode.leafMaxKeyCount() );
         this.primKeyPlaceHolder = layout.newKey();
         this.readKey = layout.newKey();
         this.readValue = layout.newValue();
+        ensureStackCapacity( 20 );
```

Comment:
Not quite a problem, it's a small "cost" of having a stack that's capable of even the craziest tree. Although we can probable set it to 10, cool?